### ISSUE TITLE
add caching to Service Worker

### ISSUE THREAD

**Issue by cvan** [mozilla/hubs#277]
this is a follow-up to issue #276.

some good examples are here:

- https://serviceworke.rs/
- https://googlechrome.github.io/samples/service-worker/

I'd recommend tackling this not any time soon but eventually. I can help advise here if there are any questions about caching strategies.

**Comment by brianpeiris** [comment]
My understanding is that Service Worker caching is primarily useful for offline apps. Obviously our app does not make sense in offline mode, so I'm not sure if this is useful for us over basic HTTP caching.

**Comment by cvan** [comment]
while Service Workers are marketed for "offline-first" web apps, Service Workers are most beneficial for reducing network time for subsequent requests ("pass-through caching" it is sometimes referred to). request, cache response, subsequent requests are fetched from the Caches bucket, the cache is updated, and so on. there are numerous strategies for Service Workers (the [Cookbook](https://serviceworke.rs) by Mozilla Dev Rel is an excellent resource for these).

what I would propose for Hubs: use a Service Worker for network resilience for repeat visits to ensure Hubs starts up more quickly when a user returns for subsequent visits. this would mean caching (and keying off content hashes) the critical-path assets: HTML, JS, CSS, images, etc.

obviously third-party services and network-dependent requests (e.g., WebSockets, WebRTC, `fetch`/XHR requests) could use the `networkFirst` caching or bypass the Service Worker's "fetch" event entirely.

**Comment by brianpeiris** [comment]
Right, but assuming we have our HTTP caching headers set correctly, we should already be performant for subsequent requests. Or is Service Worker somehow faster over good ole browser HTTP cache?

**Comment by DougReeder** [comment]
Service Workers add nothing on a first visit, but add a great deal of reliability and consistency on later visits.

More advanced uses include client-side load balancing (won't replace server-side load balancing, but can greatly help).
